anonlink/similarities/_dice.pyx

```diff
@@ -0,0 +1,186 @@
+    cdef const char[::1] one_memview = one
+    cdef const char[::1] many_memview = many
+    cdef unsigned int[::1] counts_memview = counts_many
+    cdef unsigned int[::1] indicies_memview = output_indicies
+    cdef double[::1] scores_memview = output_scores
+
+    cdef unsigned int num_elements = <unsigned int>many_memview.shape[0] // array_bytes
+
+    number_matched = c_match_one_against_many_dice_k_top(
+        &one_memview[0],
+        &many_memview[0],
+        &counts_memview[0],
+        num_elements,
+        array_bytes,
+        k,
+        threshold,
+        &indicies_memview[0],
+        &scores_memview[0]
+    )
+
+    return number_matched
+
+@cython.boundscheck(False)  # Deactivate bounds checking
+@cython.wraparound(False)   # Deactivate negative indexing.
+def dice_many_to_many(
+        const char[::1] carr0,
+        const char[::1] carr1,
+        unsigned int length_f0,
+        unsigned int length_f1,
+        unsigned int[::1] c_popcounts,
+        int filter_bytes,
+        int k,
+        double threshold,
+        array.array result_sims,
+        array.array result_indices0,
+        array.array result_indices1
+):
+    cdef size_t i
+    cdef int matches
+    cdef int total_matches = 0
+    assert len(carr1) == filter_bytes * length_f1
+    assert len(c_popcounts) == length_f1
+
+    # do all buffer allocations in Python and pass a memoryview to C
+    cdef array.array double_array_template = array.array('d', [])
+    cdef array.array int_array_template = array.array('I', [])
+
+    cdef array.array c_scores
+    cdef array.array c_indicies
+
+    c_scores = array.clone(double_array_template, k, zero=True)
+    c_indices = array.clone(int_array_template, k, zero=True)
+    i_buffer = array.clone(int_array_template, k, zero=True)
+
+    cdef double[::1] scores_memview = c_scores
+    cdef unsigned int[::1] indicies_memview = c_indices
+    cdef unsigned int[::1] i_buffer_memview = i_buffer
+
+    for i in range(length_f0):
+        #with nogil:
+        matches = match_one_to_many_dice_preallocated_output(
+            carr0[i * filter_bytes:(i + 1) * filter_bytes],
+            carr1,
+            c_popcounts,
+            length_f1,
+            filter_bytes,
+            k,
+            threshold,
+            indicies_memview,
+            scores_memview
+        )
+        total_matches += matches
+        i_buffer_memview[:] = i
+        #i_buffer_memview[:matches] = <unsigned int>i
+
+        if matches < 0:
+            raise RuntimeError('bad key length')
+
+        # Note array.extend_buffer requires the gil to extend the array
+        # `.data.as_chars` gives us direct access to the underlying contiguous C array
+        # array.extend_buffer(result_sims, c_scores.data.as_chars, matches)
+        # array.extend_buffer(result_indices0, i_buffer.data.as_chars, matches)
+        # array.extend_buffer(result_indices1, c_indices.data.as_chars, matches)
```

Comment:
what's with this section?

Reply:
Something I was trying... added more complexity and didn't allow me to make this function gil-less or faster so went back to the simpler `array.extend` method below. I'll remove the commented out code.